File: src/xai/sdk/typescript/src/clients/mining-client.ts
/**
 * Mining Client for XAI SDK
 *
 * Handles mining operations, status monitoring, and reward management.
 */

import { HTTPClient } from '../utils/http-client';
import { MiningStatus, MiningRewards } from '../types';
import { MiningError, ValidationError } from '../errors';

/**
 * Client for mining operations
 */
export class MiningClient {
  constructor(private httpClient: HTTPClient) {}

  /**
   * Start mining
   *
   * @param threads - Number of mining threads (1-16, default: 1)
   * @returns Mining status
   *
   * @example
   * ```typescript
   * const status = await client.mining.start(4);
   * console.log('Mining started with 4 threads');
   * ```
   */
  async start(threads: number = 1): Promise<Record<string, unknown>> {
    if (threads < 1 || threads > 16) {
      throw new ValidationError('threads must be between 1 and 16');
    }

    try {
      const payload = { threads };
      return await this.httpClient.post('/mining/start', payload);
    } catch (error) {
      if (error instanceof MiningError) {
        throw error;
      }
      throw new MiningError(`Failed to start mining: ${error}`);
    }
  }

  /**
   * Stop mining
   *
   * @returns Mining status
   *
   * @example
   * ```typescript
   * const status = await client.mining.stop();
   * console.log('Mining stopped');
   * ```
   */
  async stop(): Promise<Record<string, unknown>> {
    try {
      return await this.httpClient.post('/mining/stop', {});
    } catch (error) {
      if (error instanceof MiningError) {
        throw error;
      }
      throw new MiningError(`Failed to stop mining: ${error}`);
    }
  }

  /**
   * Get mining status
   *
   * @returns Mining status information
   *
   * @example
   * ```typescript
   * const status = await client.mining.getStatus();
   * console.log('Mining:', status.mining);
   * console.log('Hashrate:', status.hashrate);
   * console.log('Blocks found:', status.blocksFound);
   * ```
   */
  async getStatus(): Promise<MiningStatus> {
    try {
      const response = await this.httpClient.get<{
        mining: boolean;
        threads: number;
        hashrate: string;
        blocks_found?: number;
        current_difficulty: string;
        uptime?: number;
        last_block_time?: number;
      }>('/mining/status');

      return {
        mining: response.mining,
        threads: response.threads,
        hashrate: response.hashrate,
        blocksFound: response.blocks_found || 0,
        currentDifficulty: response.current_difficulty,
        uptime: response.uptime || 0,
        lastBlockTime: response.last_block_time,
      };
    } catch (error) {
      if (error instanceof MiningError) {
        throw error;
      }
      throw new MiningError(`Failed to get mining status: ${error}`);
    }
  }

  /**
   * Get mining rewards for an address
   *
   * @param address - Wallet address
   * @returns Reward information
   *
   * @example
   * ```typescript
   * const rewards = await client.mining.getRewards('0x1234...');
   * console.log('Total rewards:', rewards.totalRewards);
   * console.log('Pending rewards:', rewards.pendingRewards);
   * ```
   */
  async getRewards(address: string): Promise<MiningRewards> {
    if (!address) {
      throw new ValidationError('address is required');
    }

    try {
      const response = await this.httpClient.get<{
        address: string;
        total_rewards: string;
        pending_rewards: string;
        claimed_rewards: string;
        blocks_found?: number;
      }>('/mining/rewards', { address });

      return {
        address: response.address,
        totalRewards: response.total_rewards,
        pendingRewards: response.pending_rewards,
        claimedRewards: response.claimed_rewards,
        blocksFound: response.blocks_found,
      };
    } catch (error) {
      if (error instanceof MiningError) {
        throw error;
      }
      throw new MiningError(`Failed to get mining rewards: ${error}`);
    }
  }

  /**
   * Check if mining is active
   *
   * @returns True if mining is active
   *
   * @example
   * ```typescript
   * const isMining = await client.mining.isMining();
   * if (isMining) {
   *   console.log('Mining is active');
   * }
   * ```
   */
  async isMining(): Promise<boolean> {
    try {
      const status = await this.getStatus();
      return status.mining;
    } catch (error) {
      if (error instanceof MiningError) {
        throw error;
      }
      throw new MiningError(`Failed to check mining status: ${error}`);
    }
  }

  /**
   * Get current mining difficulty
   *
   * @returns Current difficulty as string
   *
   * @example
   * ```typescript
   * const difficulty = await client.mining.getDifficulty();
   * console.log('Current difficulty:', difficulty);
   * ```
   */
  async getDifficulty(): Promise<string> {
    try {
      const status = await this.getStatus();
      return status.currentDifficulty;
    } catch (error) {
      if (error instanceof MiningError) {
        throw error;
      }
      throw new MiningError(`Failed to get difficulty: ${error}`);
    }
  }

  /**
   * Get current hashrate
   *
   * @returns Current hashrate as string
   *
   * @example
   * ```typescript
   * const hashrate = await client.mining.getHashrate();
   * console.log('Current hashrate:', hashrate);
   * ```
   */
  async getHashrate(): Promise<string> {
    try {
      const status = await this.getStatus();
      return status.hashrate;
    } catch (error) {
      if (error instanceof MiningError) {
        throw error;
      }
      throw new MiningError(`Failed to get hashrate: ${error}`);
    }
  }
}

File: src/xai/sdk/python/xai_sdk/clients/mining_client.py
"""
Mining Client for XAI SDK

Handles mining operations and reward management.
"""

from typing import Optional, Dict, Any

from ..http_client import HTTPClient
from ..models import MiningStatus
from ..exceptions import MiningError, ValidationError


class MiningClient:
    """Client for mining operations."""

    def __init__(self, http_client: HTTPClient) -> None:
        """
        Initialize Mining Client.

        Args:
            http_client: HTTP client instance
        """
        self.http_client = http_client

    def start(self, threads: int = 1) -> Dict[str, Any]:
        """
        Start mining.

        Args:
            threads: Number of mining threads (1-16)

        Returns:
            Mining status

        Raises:
            MiningError: If mining start fails
        """
        if threads < 1 or threads > 16:
            raise ValidationError("threads must be between 1 and 16")

        try:
            payload = {"threads": threads}
            return self.http_client.post("/mining/start", data=payload)
        except Exception as e:
            raise MiningError(f"Failed to start mining: {str(e)}")

    def stop(self) -> Dict[str, Any]:
        """
        Stop mining.

        Returns:
            Mining status

        Raises:
            MiningError: If mining stop fails
        """
        try:
            return self.http_client.post("/mining/stop", data={})
        except Exception as e:
            raise MiningError(f"Failed to stop mining: {str(e)}")

    def get_status(self) -> MiningStatus:
        """
        Get mining status.

        Returns:
            Mining status

        Raises:
            MiningError: If status retrieval fails
        """
        try:
            response = self.http_client.get("/mining/status")

            return MiningStatus(
                mining=response["mining"],
                threads=response["threads"],
                hashrate=response["hashrate"],
                blocks_found=response.get("blocks_found", 0),
                current_difficulty=response["current_difficulty"],
                uptime=response.get("uptime", 0),
                last_block_time=response.get("last_block_time"),
            )
        except Exception as e:
            raise MiningError(f"Failed to get mining status: {str(e)}")

    def get_rewards(self, address: str) -> Dict[str, Any]:
        """
        Get mining rewards for an address.

        Args:
            address: Wallet address

        Returns:
            Reward information

        Raises:
            MiningError: If reward retrieval fails
        """
        if not address:
            raise ValidationError("address is required")

        try:
            return self.http_client.get("/mining/rewards", params={"address": address})
        except Exception as e:
            raise MiningError(f"Failed to get mining rewards: {str(e)}")

    def is_mining(self) -> bool:
        """
        Check if mining is active.

        Returns:
            True if mining is active

        Raises:
            MiningError: If check fails
        """
        try:
            status = self.get_status()
            return status.mining
        except Exception as e:
            raise MiningError(f"Failed to check mining status: {str(e)}")
